File: ecos-records/src/test/java/ru/citeck/ecos/records3/test/objdata/ObjectDataAndDataValueTest.java
package ru.citeck.ecos.records3.test.objdata;

import lombok.Data;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import ru.citeck.ecos.commons.data.DataValue;
import ru.citeck.ecos.commons.data.ObjectData;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records3.RecordsService;
import ru.citeck.ecos.records3.RecordsServiceFactory;
import ru.citeck.ecos.records3.record.op.atts.dao.RecordsAttsDao;
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao;

import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class ObjectDataAndDataValueTest extends AbstractRecordsDao implements RecordsAttsDao {

    private static final String JSON = "{\"a\":\"b\"}";

    private RecordsService recordsService;

    @NotNull
    @Override
    public String getId() {
        return "";
    }

    @BeforeAll
    void init() {
        RecordsServiceFactory factory = new RecordsServiceFactory();
        recordsService = factory.getRecordsServiceV1();
        recordsService.register(this);
    }

    @Test
    void test() {

        DataValue value = recordsService.getAtt(RecordRef.valueOf("test"), "data?json");
        assertEquals("b", value.get("a").asText());

        DataValue value2 = recordsService.getAtt(RecordRef.valueOf("test"), "data.a?str");
        assertEquals("b", value2.asText());

        value = recordsService.getAtt(RecordRef.valueOf("test"), "dataValue?json");
        assertEquals("b", value.get("a").asText());

        value2 = recordsService.getAtt(RecordRef.valueOf("test"), "dataValue.a?str");
        assertEquals("b", value2.asText());

        value2 = recordsService.getAtt(RecordRef.valueOf("test"), "data.unknown?str");
        assertEquals(DataValue.NULL, value2);

        value2 = recordsService.getAtt(RecordRef.valueOf("test"), "dataValue.unknown?str");
        assertEquals(DataValue.NULL, value2);

        value2 = recordsService.getAtt(RecordRef.valueOf("test"), "unknown?str");
        assertEquals(DataValue.NULL, value2);
    }

    @Test
    void dtoTest() {

        TestDataMeta dtoValue = recordsService.getAtts(RecordRef.valueOf("test"), TestDataMeta.class);
        TestDataMeta expected = new TestDataMeta(new TestData("test"));
        assertEquals(expected, dtoValue);
    }

    @NotNull
    @Override
    public List<?> getRecordsAtts(@NotNull List<String> records) {
        return records.stream()
            .map(TestData::new)
            .collect(Collectors.toList());
    }

    @Data
    public static class TestData {

        private ObjectData data = ObjectData.create(Json.getMapper().read(JSON, Object.class));
        private DataValue dataValue = DataValue.create(JSON);

        TestData(String ref) {
        }
    }

    @Data
    public static class TestDataMeta {

        private ObjectData data;
        private DataValue dataValue;

        public TestDataMeta(TestData data) {
            this.data = data.data;
            this.dataValue = data.dataValue;
        }

        public TestDataMeta() {
        }
    }
}

File: ecos-records/src/test/java/ru/citeck/ecos/records3/test/objdata/ObjectDataAndDataValueTest.kt
package ru.citeck.ecos.records3.test.objdata

import lombok.Data
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.data.ObjectData
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao
import ru.citeck.ecos.records3.record.dao.atts.RecordsAttsDao
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.util.stream.Collectors

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ObjectDataAndDataValueTest : AbstractRecordsDao(), RecordsAttsDao {

    companion object {
        private const val JSON = "{\"a\":\"b\"}"
    }

    override fun getId(): String {
        return "test"
    }

    @BeforeAll
    fun init() {
        val factory = RecordsServiceFactory()
        recordsService = factory.recordsService
        recordsService.register(this)
    }

    @Test
    fun testHasAtt() {

        val data = ObjectData.create()
        data.set("field", "value")

        assertThat(recordsService.getAtt(data, "_has.field?bool").asBoolean()).isTrue
        assertThat(recordsService.getAtt(data, "_has.field123?bool").asBoolean()).isFalse
    }

    @Test
    fun test() {

        val ref = EntityRef.valueOf("test@test")
        var value = recordsService.getAtt(ref, "data?json")
        assertEquals("b", value.get("a").asText())

        var value2 = recordsService.getAtt(ref, "data.a?str")
        assertEquals("b", value2.asText())

        value = recordsService.getAtt(ref, "dataValue?json")
        assertEquals("b", value.get("a").asText())

        value2 = recordsService.getAtt(ref, "dataValue.a?str")
        assertEquals("b", value2.asText())

        value2 = recordsService.getAtt(ref, "data.unknown?str")
        assertEquals(DataValue.NULL, value2)

        value2 = recordsService.getAtt(ref, "dataValue.unknown?str")
        assertEquals(DataValue.NULL, value2)

        value2 = recordsService.getAtt(ref, "unknown?str")
        assertEquals(DataValue.NULL, value2)
    }

    @Test
    fun dtoTest() {
        val dtoValue = recordsService.getAtts(EntityRef.valueOf("test@test"), TestDataMeta::class.java)
        val expected = TestDataMeta(TestData("test"))
        assertEquals(expected, dtoValue)
    }

    override fun getRecordsAtts(recordIds: List<String>): List<*> {
        return recordIds.stream()
            .map { ref: String? -> TestData(ref) }
            .collect(Collectors.toList())
    }

    @Data
    class TestData internal constructor(ref: String?) {
        val data = ObjectData.create(mapper.read(JSON, Any::class.java))
        val dataValue = DataValue.create(JSON)
    }

    data class TestDataMeta(
        var data: ObjectData? = null,
        var dataValue: DataValue? = null
    ) {
        constructor(data: TestData) : this(data.data, data.dataValue)
    }
}
